model/task/task.go

```diff
@@ -909,3 +915,68 @@ func ExpectedTaskDuration(project, buildvariant string, window time.Duration) (m
 
 	return expDurations, nil
 }
+
+// MergeNewTestResults returns the task with both old (embedded in
+// the tasks collection) and new (from the testresults collection) test results
+// merged in the Task's TestResults field.
+func (t *Task) MergeNewTestResults() error {
+	collection := Collection
+	lookupLocalField := IdKey
+	if t.Archived {
+		collection = OldCollection
+		lookupLocalField = OldTaskIdKey
+	}
+
+	// aggregation stages
+	matchStage := bson.M{"$match": bson.M{IdKey: t.Id}}
+	lookupStage := bson.M{"$lookup": bson.M{
+		"from":         testresult.Collection,
+		"localField":   lookupLocalField,
+		"foreignField": testresult.TaskIDKey,
+		"as":           newresultsAllExecutionsField,
+	}}
+	addNewResultsStage := bson.M{"$addFields": bson.M{
+		newresultsCurrentExecutionsField: bson.M{
+			"$filter": bson.M{
+				"input": "$" + newresultsAllExecutionsField,
+				"as":    "testresult",
+				"cond": bson.M{
+					"$eq": []string{
+						"$$" + "testresult" + "." + testresult.ExecutionKey,
+						"$" + ExecutionKey,
+					},
+				},
+			},
+		},
+	}}
+	concatStage := bson.M{"$addFields": bson.M{
+		TestResultsKey: bson.M{
+			"$concatArrays": []string{
+				"$" + TestResultsKey,
+				"$" + newresultsCurrentExecutionsField,
+			},
+		},
+	}}
+	cleanupStage := bson.M{"$project": bson.M{
+		newresultsAllExecutionsField:                   0,
+		newresultsCurrentExecutionsField:               0,
+		TestResultsKey + "." + IdKey:                   0,
+		TestResultsKey + "." + testresult.TaskIDKey:    0,
+		TestResultsKey + "." + testresult.ExecutionKey: 0,
+	}}
+
+	pipeline := []bson.M{
+		matchStage,
+		lookupStage,
+		addNewResultsStage,
+		concatStage,
+		cleanupStage,
+	}
+
+	tasks := []Task{}
+	if err := db.Aggregate(collection, pipeline, &tasks); err != nil {
+		return err
+	}
+	*t = tasks[0]
```

Comment:
should we include a limit 1? 

should we make sure that there is at least one matching result

Reply:
The match stage is against _id.